File: mobile/src/wirelessGecko/java/com/siliconlabs/bledemo/activity/RangeTestPresenter.java
package com.siliconlabs.bledemo.activity;

import android.os.Looper;
import android.support.annotation.UiThread;
import com.siliconlabs.bledemo.rangetest.RangeTestMode;
import com.siliconlabs.bledemo.rangetest.TxPower;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;

/**
 * @author Comarch S.A.
 */

public class RangeTestPresenter {

    public interface Controller {

        void setView(RangeTestView view);

        void initTestMode(RangeTestMode mode);

        void cancelTestMode();

        void updateTxPower(int power);

        void updatePayloadLength(int length);

        void updateMaWindowSize(int size);

        void updateChannel(int channel);

        void updatePacketCount(int count);

        void updateRemoteId(int id);

        void updateSelfId(int id);

        void updateUartLogEnabled(boolean enabled);

        void toggleRunningState();

        void updatePhyConfig(int id);
    }

    public interface RangeTestView {

        void runOnUiThread(Runnable runnable);

        @UiThread
        void showDeviceName(String name);

        @UiThread
        void showModelNumber(String number);

        @UiThread
        void showTxPower(TxPower power, List<TxPower> values);

        @UiThread
        void showPayloadLength(int length, List<Integer> values);

        @UiThread
        void showMaWindowSize(int size, List<Integer> values);

        @UiThread
        void showChannelNumber(int number);

        @UiThread
        void showPacketCountRepeat(boolean enabled);

        @UiThread
        void showPacketRequired(int required);

        @UiThread
        void showPacketSent(int sent);

        @UiThread
        void showPer(float per);

        @UiThread
        void showMa(float ma);

        @UiThread
        void showRemoteId(int id);

        @UiThread
        void showSelfId(int id);

        @UiThread
        void showUartLogEnabled(boolean enabled);

        @UiThread
        void showRunningState(boolean running);

        @UiThread
        void showTestRssi(int rssi);

        @UiThread
        void showTestRx(int received, int required);

        @UiThread
        void showPhy(int phy, LinkedHashMap<Integer, String> values);

        @UiThread
        void clearTestResults();
    }

    private static final int MA_WINDOW_MAX = 128;
    private static final int MA_WINDOW_DEFAULT = 32;

    private RangeTestView view;

    private String deviceName;
    private String modelNumber;

    private TxPower txPower;
    private List<TxPower> txPowerValues;
    private Integer payloadLength;
    private List<Integer> payloadLengthValues;
    private Integer maWindowSize;
    private List<Integer> maWindowSizeValues;

    private Integer channelNumber;
    private Boolean packetCountRepeat;
    private Integer packetRequired;
    private Integer packetReceived;
    private Integer packetSent;
    private Integer packetCount;
    private Float per;
    private Float ma;
    private Integer remoteId;
    private Integer selfId;
    private Boolean uartLogEnabled;
    private Boolean running;

    private Integer phy;
    private LinkedHashMap<Integer, String> phyMap;

    private int lastPacketCount = -1;
    private int lastPacketLoss = 0;
    private int[] maBuffer = new int[MA_WINDOW_MAX];
    private int maBufferPtr;

    @UiThread
    public void setView(RangeTestView view) {
        this.view = view;

        if (view == null) {
            return;
        }

        if (deviceName != null) view.showDeviceName(deviceName);
        if (modelNumber != null) view.showModelNumber(modelNumber);
        if (txPower != null && txPowerValues != null) view.showTxPower(txPower, txPowerValues);
        if (payloadLength != null && payloadLengthValues != null) view.showPayloadLength(payloadLength, payloadLengthValues);
        if (maWindowSize != null && maWindowSizeValues != null) view.showMaWindowSize(maWindowSize, maWindowSizeValues);
        if (channelNumber != null) view.showChannelNumber(channelNumber);
        if (packetCountRepeat != null) view.showPacketCountRepeat(packetCountRepeat);
        if (packetRequired != null) view.showPacketRequired(packetRequired);
        if (packetSent != null) view.showPacketSent(packetSent);
        if (remoteId != null) view.showRemoteId(remoteId);
        if (selfId != null) view.showSelfId(selfId);
        if (uartLogEnabled != null) view.showUartLogEnabled(uartLogEnabled);
        if (running != null) view.showRunningState(running);
        if (phy != null && phyMap != null) view.showPhy(phy, phyMap);

        view.showMa(ma == null ? 0 : ma);
        view.showPer(per == null ? 0 : per);

        if (packetReceived == null || packetCount == null) {
            view.showTestRx(0, 0);
        } else {
            view.showTestRx(packetReceived, packetCount);
        }
    }

    // controller -> view

    void onDeviceNameUpdated(final String deviceName) {
        this.deviceName = deviceName;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showDeviceName(deviceName);
            }
        });
    }

    void onModelNumberUpdated(final String number) {
        this.modelNumber = number;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showModelNumber(number);
            }
        });
    }

    void onTxPowerUpdated(final int power) {
        this.txPower = new TxPower(power);
        if (txPowerValues != null) {
            onView(new ViewAction() {
                @Override
                public void run(RangeTestView view) {
                    view.showTxPower(txPower, txPowerValues);
                }
            });
        }
    }

    void onPayloadLengthUpdated(final int length) {
        this.payloadLength = length;
        if (payloadLengthValues != null) {
            onView(new ViewAction() {
                @Override
                public void run(RangeTestView view) {
                    view.showPayloadLength(length, payloadLengthValues);
                }
            });
        }
    }

    void onMaWindowSizeUpdated(final int size) {
        this.maWindowSize = size;
        if (maWindowSizeValues != null) {
            onView(new ViewAction() {
                @Override
                public void run(RangeTestView view) {
                    view.showMaWindowSize(size, maWindowSizeValues);
                }
            });
        }
    }

    void onChannelNumberUpdated(final int number) {
        this.channelNumber = number;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showChannelNumber(number);
            }
        });
    }

    void onPacketCountRepeatUpdated(final boolean enabled) {
        this.packetCountRepeat = enabled;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showPacketCountRepeat(enabled);
            }
        });
    }

    void onPacketRequiredUpdated(final int required) {
        this.packetRequired = required;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showPacketRequired(required);
            }
        });
    }

    void onPacketReceivedUpdated(final int received) {
        this.packetReceived = received;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                if (RangeTestPresenter.this.packetCount != null) {
                    view.showTestRx(received, RangeTestPresenter.this.packetCount);
                }
            }
        });
    }

    void onPacketSentUpdated(final int sent) {
        if (sent <= packetRequired) {
            this.packetSent = sent;
            onView(new ViewAction() {
                @Override
                public void run(RangeTestView view) {
                    view.showPacketSent(sent);
                }
            });
        }
    }

    void onPacketCountUpdated(final int packetCount) {
        this.packetCount = packetCount;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                if (RangeTestPresenter.this.packetReceived != null) {
                    view.showTestRx(RangeTestPresenter.this.packetReceived, packetCount);
                }
            }
        });
    }

    void onMaUpdated(final float ma) {
        this.ma = ma;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showMa(ma);
            }
        });
    }

    void onPerUpdated(final float per) {
        this.per = per;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showPer(per);
            }
        });
    }

    void onRemoteIdUpdated(final int id) {
        this.remoteId = id;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showRemoteId(id);
            }
        });
    }

    void onSelfIdUpdated(final int id) {
        this.selfId = id;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showSelfId(id);
            }
        });
    }

    void onUartLogEnabledUpdated(final boolean enabled) {
        this.uartLogEnabled = enabled;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showUartLogEnabled(enabled);
            }
        });
    }

    void onRunningStateUpdated(final boolean running) {
        this.running = running;
        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showRunningState(running);
            }
        });
    }

    void onPhyConfigUpdated(final int phy) {
        this.phy = phy;
        if (phyMap != null) {
            onView(new ViewAction() {
                @Override
                public void run(RangeTestView view) {
                    view.showPhy(phy, phyMap);
                }
            });
        }
    }

    void onPhyMapUpdated(final LinkedHashMap<Integer, String> phyMap) {
        this.phyMap = phyMap;
        if (phy != null) {
            onView(new ViewAction() {
                @Override
                public void run(RangeTestView view) {
                    view.showPhy(phy, phyMap);
                }
            });
        }
    }

    void onTxPowerRangeUpdated(final int from, final int to) {
        List<TxPower> values = new ArrayList<>();
        for (int value = from; value <= to; value = value + 5) {
            values.add(new TxPower(value));
        }
        txPowerValues = values;

        if (txPower != null) {
            onView(new ViewAction() {
                @Override
                public void run(RangeTestView view) {
                    view.showTxPower(txPower, txPowerValues);
                }
            });
        }
    }

    void onPayloadLengthRangeUpdated(final int from, final int to) {
        List<Integer> values = new ArrayList<>();
        for (int i = 0; i < to - from + 1; ++i) {
            values.add(from + i);
        }
        payloadLengthValues = values;

        if (payloadLength != null) {
            onView(new ViewAction() {
                @Override
                public void run(RangeTestView view) {
                    view.showPayloadLength(payloadLength, payloadLengthValues);
                }
            });
        }
    }

    void onMaWindowSizeRangeUpdated(final int from, final int to) {
        List<Integer> values = new ArrayList<>();
        for (int i = from; i <= to; i = i * 2) {
            values.add(i);
        }
        maWindowSizeValues = values;

        if (maWindowSize != null) {
            onView(new ViewAction() {
                @Override
                public void run(RangeTestView view) {
                    view.showMaWindowSize(maWindowSize, maWindowSizeValues);
                }
            });
        }
    }

    void onTestDataReceived(final int rssi, final int packetCount, final int packetReceived) {
        if (packetCount < this.lastPacketCount) {
            clearMaBuffer();
            lastPacketLoss = 0;

            onView(new ViewAction() {
                @Override
                public void run(RangeTestView view) {
                    view.showPer(0);
                    view.showMa(0);
                    view.clearTestResults();
                }
            });
        }

        this.lastPacketCount = packetCount;

//        this.packetRequired = packetCount;
        this.packetReceived = packetReceived;

        int totalPacketLoss = packetCount - packetReceived;
        int currentPacketLoss = totalPacketLoss - lastPacketLoss;
        lastPacketLoss = totalPacketLoss;

        final float per = totalPacketLoss * 100f / (float) packetCount;
        final float ma = updateMa(currentPacketLoss);

        onView(new ViewAction() {
            @Override
            public void run(RangeTestView view) {
                view.showTestRx(packetReceived, packetCount);
                view.showPer(per);
                view.showMa(ma);
                view.showTestRssi(rssi);
            }
        });
    }

    private void onView(final ViewAction action) {
        final RangeTestView view = this.view;

        if (view == null) {
            return;
        }

        if (Thread.currentThread() == Looper.getMainLooper().getThread()) {
            action.run(view);
        } else {
            view.runOnUiThread(new Runnable() {

                @Override
                public void run() {
                    action.run(view);
                }
            });
        }
    }

    private float updateMa(int packetsLost) {
        maBuffer[maBufferPtr++] = packetsLost;

        if (maBufferPtr >= maBuffer.length) {
            maBufferPtr = 0;
        }

        int window = getMaWindow();
        return sumMaBuffer(window) * 100f / (float) window;
    }

    private int sumMaBuffer(int window) {
        int sum = 0;

        for (int i = 1; i <= window; ++i) {
            int location = maBufferPtr - i;
            if (location < 0) {
                location = maBuffer.length - i;
            }

            int loss = maBuffer[location];
            if (loss >= 0) {
                sum += loss;
            } else {
                break;
            }
        }

        return sum;
    }

    private int getMaWindow() {
        Integer window = maWindowSize;
        return window != null && window > 0 && window <= MA_WINDOW_MAX ? window : MA_WINDOW_DEFAULT;
    }

    private void clearMaBuffer() {
        Arrays.fill(maBuffer, -1);
        maBufferPtr = 0;
    }

    private interface ViewAction {

        void run(RangeTestView view);
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/RangeTest/Presenters/RangeTestPresenter.kt
package com.siliconlabs.bledemo.RangeTest.Presenters

import android.os.Looper
import androidx.annotation.UiThread
import com.siliconlabs.bledemo.Bluetooth.BLE.BluetoothDeviceInfo
import com.siliconlabs.bledemo.RangeTest.Models.DeviceState
import com.siliconlabs.bledemo.RangeTest.Models.RangeTestMode
import com.siliconlabs.bledemo.RangeTest.Models.TxPower
import java.util.*

/**
 * @author Comarch S.A.
 */
class RangeTestPresenter {

    interface Controller {
        fun setView(view: RangeTestView?)
        fun initTestMode(mode: RangeTestMode?)
        fun cancelTestMode()
        fun updateTxPower(power: Int)
        fun updatePayloadLength(length: Int)
        fun updateMaWindowSize(size: Int)
        fun updateChannel(channel: Int)
        fun updatePacketCount(count: Int)
        fun updateRemoteId(id: Int)
        fun updateSelfId(id: Int)
        fun updateUartLogEnabled(enabled: Boolean)
        fun toggleRunningState()
        fun updatePhyConfig(id: Int)
    }

    interface RangeTestView {
        fun runOnUiThread(runnable: Runnable?)

        @UiThread
        fun showDeviceName(name: String?)

        @UiThread
        fun showModelNumber(number: String?, running: Boolean?)

        @UiThread
        fun showTxPower(power: TxPower?, values: List<TxPower>)

        @UiThread
        fun showPayloadLength(length: Int, values: List<Int>)

        @UiThread
        fun showMaWindowSize(size: Int, values: List<Int>)

        @UiThread
        fun showChannelNumber(number: Int)

        @UiThread
        fun showPacketCountRepeat(enabled: Boolean)

        @UiThread
        fun showPacketRequired(required: Int)

        @UiThread
        fun showPacketSent(sent: Int)

        @UiThread
        fun showPer(per: Float)

        @UiThread
        fun showMa(ma: Float)

        @UiThread
        fun showRemoteId(id: Int)

        @UiThread
        fun showSelfId(id: Int)

        @UiThread
        fun showUartLogEnabled(enabled: Boolean)

        @UiThread
        fun showRunningState(running: Boolean)

        @UiThread
        fun showTestRssi(rssi: Int)

        @UiThread
        fun showTestRx(received: Int, required: Int)

        @UiThread
        fun showPhy(phy: Int, values: LinkedHashMap<Int, String>)

        @UiThread
        fun clearTestResults()
    }

    var deviceState1 = DeviceState()
    var deviceState2 = DeviceState()
    var currentDevice = deviceState1

    private var view: RangeTestView? = null

    @UiThread
    fun setView(view: RangeTestView?) {
        this.view = view

        if (view == null) {
            return
        }

        if (currentDevice.deviceName != null) view.showDeviceName(currentDevice.deviceName)
        if (currentDevice.modelNumber != null) view.showModelNumber(currentDevice.modelNumber, currentDevice.running)
        if (currentDevice.txPower != null && currentDevice.txPowerValues != null) view.showTxPower(currentDevice.txPower, currentDevice.txPowerValues!!)
        if (currentDevice.payloadLength != null && currentDevice.payloadLengthValues != null) view.showPayloadLength(currentDevice.payloadLength!!, currentDevice.payloadLengthValues!!)
        if (currentDevice.maWindowSize != null && currentDevice.maWindowSizeValues != null) view.showMaWindowSize(currentDevice.maWindowSize!!, currentDevice.maWindowSizeValues!!)
        if (currentDevice.channelNumber != null) view.showChannelNumber(currentDevice.channelNumber!!)
        if (currentDevice.packetCountRepeat != null) view.showPacketCountRepeat(currentDevice.packetCountRepeat!!)
        if (currentDevice.packetsRequired != null) view.showPacketRequired(currentDevice.packetsRequired!!)
        if (currentDevice.packetsSent != null) view.showPacketSent(currentDevice.packetsSent!!)
        if (currentDevice.remoteId != null) view.showRemoteId(currentDevice.remoteId!!)
        if (currentDevice.selfId != null) view.showSelfId(currentDevice.selfId!!)
        if (currentDevice.uartLogEnabled != null) view.showUartLogEnabled(currentDevice.uartLogEnabled!!)
        if (currentDevice.running != null) view.showRunningState(currentDevice.running!!)
        if (currentDevice.phy != null && currentDevice.phyMap != null) view.showPhy(currentDevice.phy!!, currentDevice.phyMap!!)
        view.showMa(if (currentDevice.ma == null) 0f else currentDevice.ma!!)
        view.showPer(if (currentDevice.per == null) 0f else currentDevice.per!!)
        if (currentDevice.packetsReceived == null || currentDevice.packetCount == null) {
            view.showTestRx(0, 0)
        } else {
            view.showTestRx(currentDevice.packetsReceived!!, currentDevice.packetCount!!)
        }
    }

    fun getDeviceByAddress(address: String) : DeviceState? {
        return when (address) {
            deviceState1.deviceInfo?.address -> deviceState1
            deviceState2.deviceInfo?.address -> deviceState2
            else -> null
        }
    }

    fun getDeviceInfo()  = currentDevice.deviceInfo
    fun getDeviceInfoAt(which: Int) : BluetoothDeviceInfo? {
        return if (which == 1) deviceState1.deviceInfo
               else deviceState2.deviceInfo
    }
    fun setDeviceInfo(info: BluetoothDeviceInfo?) {
        currentDevice.deviceInfo = info
    }
    fun setDeviceInfoAt(which: Int, info: BluetoothDeviceInfo?) {
        if (which == 1) deviceState1.deviceInfo = info
        else deviceState2.deviceInfo = info
    }

    fun getMode() = currentDevice.mode
    fun setMode(mode: RangeTestMode?) {
        currentDevice.mode = mode
    }
    fun setModeAt(which: Int, mode: RangeTestMode?) {
        if (which == 1) deviceState1.mode = mode
        else deviceState2.mode = mode
    }

    fun getTestRunning() = currentDevice.testRunning
    fun setTestRunning(address: String, running: Boolean) {
        getDeviceByAddress(address)?.testRunning = running
    }
    fun setTestRunningAt(which: Int, running: Boolean) {
        if (which == 1) deviceState1.testRunning = running
        else deviceState2.testRunning = running
    }

    fun resetDeviceAt(which: Int) {
        setDeviceInfoAt(which, null)
        setModeAt(which, null)
        setTestRunningAt(which, false)
    }

    fun switchCurrentDevice() {
        currentDevice =
                if (currentDevice === deviceState1) deviceState2
                else deviceState1
    }

    fun setCurrentDevice(which: Int) {
        currentDevice =
                if (which == 1) deviceState1
                else deviceState2
    }

    // controller -> view
    fun onDeviceNameUpdated(address: String, deviceName: String?) {
        getDeviceByAddress(address)?.let {
            it.deviceName = deviceName
            onView(it, object : ViewAction {
                override fun run(view: RangeTestView) {
                    view.showDeviceName(deviceName)
                }
            })
        }

    }

    fun onModelNumberUpdated(address: String, number: String?) {
        getDeviceByAddress(address)?.let {
            it.modelNumber = number
            onView(it, object : ViewAction {
                override fun run(view: RangeTestView) {
                    view.showModelNumber(number, it.running)
                }
            })
        }

    }

    fun onTxPowerUpdated(address: String, power: Int) {
        getDeviceByAddress(address)?.let {
            it.txPower = TxPower(power)
            if (it.txPowerValues != null) {
                onView(it, object : ViewAction {
                    override fun run(view: RangeTestView) {
                        view.showTxPower(it.txPower, it.txPowerValues!!)
                    }
                })
            }
        }
    }

    fun onPayloadLengthUpdated(address: String, length: Int) {
        getDeviceByAddress(address)?.let {
            it.payloadLength = length
            if (it.payloadLengthValues != null) {
                onView(it, object : ViewAction {
                    override fun run(view: RangeTestView) {
                        view.showPayloadLength(length, it.payloadLengthValues!!)
                    }
                })
            }
        }

    }

    fun onMaWindowSizeUpdated(address: String, size: Int) {
        getDeviceByAddress(address)?.let {
            it.maWindowSize = size
            if (it.maWindowSizeValues != null) {
                onView(it, object : ViewAction {
                    override fun run(view: RangeTestView) {
                        view.showMaWindowSize(size, it.maWindowSizeValues!!)
                    }
                })
            }
        }
    }

    fun onChannelNumberUpdated(address: String, number: Int) {
        getDeviceByAddress(address)?.let {
            it.channelNumber = number
            onView(it, object : ViewAction {
                override fun run(view: RangeTestView) {
                    view.showChannelNumber(number)
                }
            })
        }
    }

    fun onPacketCountRepeatUpdated(address: String, enabled: Boolean) {
        getDeviceByAddress(address)?.let {
            it.packetCountRepeat = enabled
            onView(it, object : ViewAction {
                override fun run(view: RangeTestView) {
                    view.showPacketCountRepeat(enabled)
                }
            })
        }
    }

    fun onPacketRequiredUpdated(address: String, required: Int) {
        getDeviceByAddress(address)?.let {
            it.packetsRequired = required
            onView(it, object : ViewAction {
                override fun run(view: RangeTestView) {
                    view.showPacketRequired(required)
                }
            })
        }
    }

    fun onPacketSentUpdated(address: String, sent: Int) {
        getDeviceByAddress(address)?.let {
            it.packetsRequired?.let { packetRequired ->
                if (sent <= packetRequired) {
                    it.packetsSent = sent
                    onView(it, object : ViewAction {
                        override fun run(view: RangeTestView) {
                            view.showPacketSent(sent)
                        }
                    })
                }
            }
        }
    }

    fun onRemoteIdUpdated(address: String, id: Int) {
        getDeviceByAddress(address)?.let {
            it.remoteId = id
            onView(it, object : ViewAction {
                override fun run(view: RangeTestView) {
                    view.showRemoteId(id)
                }
            })
        }
    }

    fun onSelfIdUpdated(address: String, id: Int) {
        getDeviceByAddress(address)?.let {
            it.selfId = id
            onView(it, object : ViewAction {
                override fun run(view: RangeTestView) {
                    view.showSelfId(id)
                }
            })
        }
    }

    fun onUartLogEnabledUpdated(address: String, enabled: Boolean) {
        getDeviceByAddress(address)?.let {
            it.uartLogEnabled = enabled
            onView(it, object : ViewAction {
                override fun run(view: RangeTestView) {
                    view.showUartLogEnabled(enabled)
                }
            })
        }
    }

    fun onRunningStateUpdated(address: String, running: Boolean) {
        getDeviceByAddress(address)?.let {
            it.running = running
            onView(it, object : ViewAction {
                override fun run(view: RangeTestView) {
                    view.showRunningState(running)
                }
            })
        }
    }

    fun onPhyConfigUpdated(address: String, phy: Int) {
        getDeviceByAddress(address)?.let {
            it.phy = phy
            if (it.phyMap != null) {
                onView(it, object : ViewAction {
                    override fun run(view: RangeTestView) {
                        view.showPhy(phy, it.phyMap!!)
                    }
                })
            }
        }
    }

    fun onPhyMapUpdated(address: String, phyMap: LinkedHashMap<Int, String>) {
        getDeviceByAddress(address)?.let {
            it.phyMap = phyMap
            if (it.phy != null) {
                onView(it, object : ViewAction {
                    override fun run(view: RangeTestView) {
                        view.showPhy(it.phy!!, phyMap)
                    }
                })
            }
        }
    }

    fun onTxPowerRangeUpdated(address: String, from: Int, to: Int) {
        val values: MutableList<TxPower> = ArrayList()
        var value = from
        while (value <= to) {
            values.add(TxPower(value))
            value += 5
        }

        getDeviceByAddress(address)?.let {
            it.txPowerValues = values
            if (it.txPower != null) {
                onView(it, object : ViewAction {
                    override fun run(view: RangeTestView) {
                        view.showTxPower(it.txPower, it.txPowerValues!!)
                    }
                })
            }
        }
    }

    fun onPayloadLengthRangeUpdated(address: String, from: Int, to: Int) {
        val values: MutableList<Int> = ArrayList()
        for (i in 0 until to - from + 1) {
            values.add(from + i)
        }

        getDeviceByAddress(address)?.let {
            it.payloadLengthValues = values
            if (it.payloadLength != null) {
                onView(it, object : ViewAction {
                    override fun run(view: RangeTestView) {
                        view.showPayloadLength(it.payloadLength!!, it.payloadLengthValues!!)
                    }
                })
            }
        }
    }

    fun onMaWindowSizeRangeUpdated(address: String, from: Int, to: Int) {
        val values: MutableList<Int> = ArrayList()
        var i = from
        while (i <= to) {
            values.add(i)
            i *= 2
        }

        getDeviceByAddress(address)?.let {
            it.maWindowSizeValues = values
            if (it.maWindowSize != null) {
                onView(it, object : ViewAction {
                    override fun run(view: RangeTestView) {
                        view.showMaWindowSize(it.maWindowSize!!, it.maWindowSizeValues!!)
                    }
                })
            }
        }
    }

    fun onTestDataReceived(address: String, rssi: Int, packetCount: Int, packetReceived: Int) {
        getDeviceByAddress(address)?.let {
            if (packetCount < it.lastPacketCount) {
                clearMaBuffer(it)
                it.lastPacketLoss = 0
                onView(it, object : ViewAction {
                    override fun run(view: RangeTestView) {
                        view.showPer(0f)
                        view.showMa(0f)
                        view.clearTestResults()
                    }
                })
            }
            it.lastPacketCount = packetCount

            //        this.packetRequired = packetCount;
            it.packetsReceived = packetReceived
            val totalPacketLoss = packetCount - packetReceived
            val currentPacketLoss = totalPacketLoss - it.lastPacketLoss
            it.lastPacketLoss = totalPacketLoss
            val per = totalPacketLoss * 100f / packetCount.toFloat()
            val ma = updateMa(it, currentPacketLoss)
            onView(it, object : ViewAction {
                override fun run(view: RangeTestView) {
                    view.showTestRx(packetReceived, packetCount)
                    view.showPer(per)
                    view.showMa(ma)
                    view.showTestRssi(rssi)
                }
            })
        }
    }

    private fun onView(deviceState: DeviceState, action: ViewAction) {
        if (deviceState === currentDevice) {
            val view = view ?: return
            if (Thread.currentThread() === Looper.getMainLooper().thread) {
                action.run(view)
            } else {
                view.runOnUiThread(Runnable { action.run(view) })
            }
        }
    }

    private fun updateMa(deviceState: DeviceState, packetsLost: Int): Float {
        deviceState.maBuffer[deviceState.maBufferPtr++] = packetsLost

        if (deviceState.maBufferPtr >= deviceState.maBuffer.size) {
            deviceState.maBufferPtr = 0
        }

        val window = getMaWindow(deviceState)
        return sumMaBuffer(deviceState, window) * 100f / window.toFloat()
    }

    private fun sumMaBuffer(deviceState: DeviceState, window: Int): Int {
        var sum = 0
        for (i in 1..window) {
            var location = deviceState.maBufferPtr - i
            if (location < 0) {
                location = deviceState.maBuffer.size - i
            }
            val loss = deviceState.maBuffer[location]
            sum += if (loss >= 0) {
                loss
            } else {
                break
            }
        }
        return sum
    }

    private fun getMaWindow(deviceState: DeviceState) : Int {
        val window = deviceState.maWindowSize
        return if (window != null && window > 0 && window <= MA_WINDOW_MAX) window else MA_WINDOW_DEFAULT
    }

    private fun clearMaBuffer(deviceState: DeviceState) {
        Arrays.fill(deviceState.maBuffer, -1)
        deviceState.maBufferPtr = 0
    }

    private interface ViewAction {
        fun run(view: RangeTestView)
    }

    companion object {
        private const val MA_WINDOW_MAX = 128
        private const val MA_WINDOW_DEFAULT = 32
    }
}